### DISCUSSION TITLE
BeautifulSoup-網路爬蟲的重要拼圖

### DISCUSSION THREAD

**Discussion by eric202501** [1101-GS4719-C/Python-Community-GS4719-C#1602] [優秀Python套件]
BeautifulSoup 是一個 Python 的函式庫，有觀察網路爬蟲這方面可能有看過這套件，它可以從 HTML 或 XML 檔案中分析資料，並將我們的目標資料抓出來分析，並將分析的結果轉換成 tag ，一般常用在網路探勘和資料蒐集上。

使用教學
在CMD中輸入pip install beautifulsoup4

簡單的示範
import requests
from bs4 import BeautifulSoup

url = 'https://www.google.com.tw/?hl=zh_TW'
web = requests.get(url)  # 取得網頁內容
soup = BeautifulSoup(web.text, "html.parser")  # 轉換成標籤樹
title = soup.title  # 取得 title
print(title)  

BeautifulSoup 定位的函式
soup.find() : 根據條件回傳"第一個"符合的元素，由字串表示，若沒有符合的則回傳 None。
soup.find_all() : 根據條件回傳"全部"符合的元素，由串列表示，若沒有符合的則回傳空串列。
soup.select() : Css Selector。
可以透過標籤、 id 或 class 來定位元素， soup.find('p', id='myid', class_='myclass') ，這裡的 class 後方必須加上底線，為了避免和Python 的關鍵字 class 衝突。

更多詳細的有關教學可以從 https://steam.oxxostudio.tw/category/python/spider/beautiful-soup.html
感謝閱讀

**Comment by v0ouo** [comment]
看完後覺得好複雜，就算點進去連結後也還是不太懂BeautifulSoup函式庫的含意，可以請教本人關於這篇文的問題嗎

**Reply by eric202501** [reply]
你可以晚上來我房間 我們可以好好討論一下

**Reply by v0ouo** [reply]
好啊，房間在哪裡呢，要順便請我吃宵夜嗎?

**Reply by mrweixd** [reply]
不糾，我也想加入討論。

**Comment by mrweixd** [comment]
想問為什麼為什麼這個套件叫做BeautifulSoup呢？美麗的湯聽起來蠻有趣的

**Reply by eric202501** [reply]
Beautiful Soup 函式庫是以愛麗絲夢遊仙境中的一首詩命名
故事中此詩由 Mock Turtle 吟誦
如同夢遊仙境，Beautiful Soup 試著讓無厘頭有道理
它矯正不良 HTML 的一團亂以產生 XML 結構的 Python 物件 
我爬巴哈找到的

**Comment by lazywulf** [comment]
BeautifulSoup 的文檔寫的比隔壁棚 Scrapy 好很多。要快速爬資料非常好用 (雖說要先熟悉 Python)。
但有時候遇到網站 API 寫很好的好像就不大用了，像是某些網站會直接提供 JSON 格式等等。他也只不過是 parser。我覺得重點應該擺在如何高速收發 request。Blocking 跟 non-blocking I/O 其實差別很大，幾千張圖片可能是好幾十分鐘的差距。另外還要注意如何不讓對方伺服器以為你在搞 DoS 而被擋下來。

要注意宿網用量喔。發 request 會吃流量喔。吃爆你加油。
